crates/proof-of-sql/src/sql/postprocessing/group_by_postprocessing.rs

```diff
@@ -0,0 +1,325 @@
+/// Get identifiers NOT in aggregate functions
+fn get_free_identifiers_from_expr(expr: &Expression) -> IndexSet<Identifier> {
+    match expr {
+        Expression::Column(identifier) => IndexSet::from([*identifier]),
+        Expression::Literal(_) | Expression::Aggregation { .. } | Expression::Wildcard => {
+            IndexSet::new()
+        }
+        Expression::Binary { left, right, .. } => {
+            let left_identifiers = get_free_identifiers_from_expr(left);
+            let right_identifiers = get_free_identifiers_from_expr(right);
+            left_identifiers
+                .union(&right_identifiers)
+                .cloned()
+                .collect::<IndexSet<_>>()
+        }
+        Expression::Unary { expr, .. } => get_free_identifiers_from_expr(expr),
+    }
+}
+
+/// Get aggregate expressions from an expression as well as the remainder
+///
+/// The idea here is to recursively traverse the expression tree and collect all the aggregation expressions
+/// and then label them as new columns post-aggregation and replace them with these new columns so that
+/// the post-aggregation expression tree doesn't contain any aggregation expressions and can be simply evaluated.
+fn get_aggregate_and_remainder_expressions(
+    expr: &Expression,
+    aggregation_expr_map: &mut IndexMap<(AggregationOperator, Expression), Identifier>,
+) -> Expression {
+    match expr {
+        Expression::Column(_) | Expression::Literal(_) | Expression::Wildcard => expr.clone(),
+        Expression::Aggregation { op, expr } => {
+            let key = (*op, *(*expr).clone());
+            if !aggregation_expr_map.contains_key(&key) {
+                let new_col_id = format!("__col_agg_{}", aggregation_expr_map.len())
+                    .parse()
+                    .unwrap();
+                aggregation_expr_map.insert(key, new_col_id);
+                Expression::Column(new_col_id)
+            } else {
+                Expression::Column(*aggregation_expr_map.get(&key).unwrap())
+            }
+        }
+        Expression::Binary { op, left, right } => {
+            let left_remainder =
+                get_aggregate_and_remainder_expressions(left, aggregation_expr_map);
+            let right_remainder =
+                get_aggregate_and_remainder_expressions(right, aggregation_expr_map);
+            Expression::Binary {
+                op: *op,
+                left: Box::new(left_remainder),
+                right: Box::new(right_remainder),
+            }
+        }
+        Expression::Unary { op, expr } => {
+            let remainder = get_aggregate_and_remainder_expressions(expr, aggregation_expr_map);
+            Expression::Unary {
+                op: *op,
+                expr: Box::new(remainder),
+            }
+        }
+    }
+}
+
+/// Given an `AliasedResultExpr`, check if it is legitimate and if so grab the relevant aggregation expression
+fn check_and_get_aggregation_and_remainder(
+    expr: &AliasedResultExpr,
```

Comment:
See other ⭐ comments.
```suggestion
    expr: AliasedResultExpr,
```